std/shared/src/main/scala/cats/effect/std/NonEmptyHotswap.scala

```diff
@@ -0,0 +1,153 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect.std
+
+import cats.effect.kernel.{Concurrent, Ref, Resource}
+import cats.syntax.all._
+
+/**
+ * A concurrent data structure that exposes a linear sequence of `R` resources as a single
+ * [[cats.effect.kernel.Resource]] in `F` without accumulation.
+ *
+ * A [[NonEmptyHotswap]] is allocated within a [[cats.effect.kernel.Resource]] that dictates the
+ * scope of its lifetime. After creation, a `Resource[F, R]` can be swapped in by calling
+ * [[swap]]. The newly acquired resource is returned and is released either when the
+ * [[NonEmptyHotswap]] is finalized or upon the next call to [[swap]], whichever occurs first.
+ *
+ * The following diagram illustrates the linear allocation and release of three resources `r1`,
+ * `r2`, and `r3` cycled through [[NonEmptyHotswap]]:
+ *
+ * {{{
+ * create(r1) ----- swap(r2) ---- swap(r3) ---- X
+ * |                |             |             |
+ * r1 acquired      |             |             |
+ *                  r2 acquired   |             |
+ *                  r1 released   r3 acquired   |
+ *                                r2 released   |
+ *                                              r3 released
+ * }}}
+ *
+ * [[NonEmptyHotswap]] is particularly useful when working with effects that cycle through
+ * resources, like writing bytes to files or rotating files every N bytes or M seconds. Without
+ * [[NonEmptyHotswap]], such effects leak resources: on each file rotation, a file handle or
+ * some internal resource handle accumulates. With [[NonEmptyHotswap]], the only registered
+ * resource is the [[NonEmptyHotswap]] itself, and each file is swapped in only after swapping
+ * the previous one out.
+ *
+ * Born from [[Hotswap]].
+ */
+sealed trait NonEmptyHotswap[F[_], R] {
+
+  /**
+   * Allocates a new resource, closes the previous one, and returns the newly allocated `R`.
+   *
+   * When the lifetime of the [[NonEmptyHotswap]] is completed, the resource allocated by the
+   * most recent [[swap]] will be finalized.
+   *
+   * [[swap]] finalizes the previous resource immediately, so users must ensure that the old `R`
+   * is not used thereafter. Failure to do so may result in an error on the _consumer_ side. In
+   * any case, no resources will be leaked.
+   *
+   * For safer access to the current resource see [[get]], which guarantees that it will not be
+   * released while it is being used.
+   *
+   * If [[swap]] is called after the lifetime of the [[NonEmptyHotswap]] is over, it will raise
+   * an error, but will ensure that all resources are finalized before returning.
+   */
+  def swap(next: Resource[F, R]): F[R]
```

Comment:
This signature is actually unsafe, because you should never return `R` without acquiring the semaphore lock.

It must be either:

```suggestion
  def swap(next: Resource[F, R]): F[Unit]
```
or:
```suggestion
  def swap(next: Resource[F, R]): Resource[F, R]
```

Reply:
I'm happy with `hs.swap(...) *> hs.get`